File: leetcode/42.接雨水.ts
/*
 * @lc app=leetcode.cn id=42 lang=typescript
 *
 * [42] 接雨水
 */

// @lc code=start
function trap(height: number[]): number {
    let ret = 0;
    let [left, right] = [0, height.length - 1];
    let [leftMax, rightMax] = [0, 0];
    while (left < right) {
        leftMax = Math.max(leftMax, height[left]);
        rightMax = Math.max(rightMax, height[right]);
        if (height[left] < height[right]) {
            ret += leftMax - height[left];
            left++;
        } else {
            ret += rightMax - height[right];
            right--;
        }
    }
    return ret;
};
// @lc code=end

File: leetcode/42.接雨水.rs
/*
 * @lc app=leetcode.cn id=42 lang=rust
 *
 * [42] 接雨水
 */

// @lc code=start
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let (mut left, mut right) = (0, height.len() - 1);
        let (mut left_max, mut right_max) = (0, 0);
        let mut ret = 0;
        while left < right {
            left_max = left_max.max(height[left]);
            right_max = right_max.max(height[right]);
            
            if height[left] < height[right] {
                ret += left_max - height[left];
                left += 1;
            } else {
                ret += right_max - height[right];
                right -= 1;
            }
        }
        ret
    }
}
// @lc code=end

